packages/backend/src/core/ValidiumSpotDataSyncService.ts

```diff
@@ -0,0 +1,173 @@
+import { AvailabilityGatewayClient } from '../peripherals/starkware/AvailabilityGatewayClient'
+import { PerpetualBatch } from '../peripherals/starkware/toPerpetualBatch'
+import { Logger } from '../tools/Logger'
+import { LogRootUpdate, LogStateTransitionFact } from './collectors/events'
+import { FinalizeExitEventsCollector } from './collectors/FinalizeExitEventsCollector'
+import { ForcedEventsCollector } from './collectors/ForcedEventsCollector'
+import { UserRegistrationCollector } from './collectors/UserRegistrationCollector'
+import { StateUpdater } from './StateUpdater'
+
+interface ValidiumStateTransition {
+  blockNumber: number
+  transactionHash: Hash256
+  stateTransitionFact: Hash256
+  sequenceNumber: number
+  batchId: number
+}
+
+export class ValidiumSpotDataSyncService {
+  constructor(
+    private readonly ethereumClient: EthereumClient,
+    private readonly availabilityGatewayClient: AvailabilityGatewayClient,
+    private readonly perpetual: EthereumAddress,
+    private readonly stateUpdater: StateUpdater,
+    private readonly userRegistrationCollector: UserRegistrationCollector,
+    private readonly forcedEventsCollector: ForcedEventsCollector,
+    private readonly finalizeExitEventsCollector: FinalizeExitEventsCollector,
+    private readonly logger: Logger
+  ) {
+    this.logger = logger.for(this)
+  }
+
+  async sync(blockRange: BlockRange) {
+    const userRegistrations = await this.userRegistrationCollector.collect(
+      blockRange
+    )
+    const forcedEvents = await this.forcedEventsCollector.collect(blockRange)
+
+    const finalizeExitEvents = await this.finalizeExitEventsCollector.collect(
+      blockRange
+    )
+
+    const stateTransitions = await this.collectValidiumStateTransitions(
+      blockRange
+    )
+
+    this.logger.info({
+      method: 'validium sync',
+      blockRange: { from: blockRange.start, to: blockRange.end },
+      stateTransitions: stateTransitions.length,
+      userRegistrations: userRegistrations.length,
+      forcedEvents,
+      finalizeExitEvents,
+    })
+
+    await this.processValidiumStateTransitions(stateTransitions)
+  }
+
+  private async processValidiumStateTransitions(
+    validiumStateTransition: ValidiumStateTransition[]
+  ) {
+    const { oldHash, id } = await this.stateUpdater.readLastUpdate()
+    await this.stateUpdater.ensureRollupState(oldHash)
+
+    for (const [i, transition] of validiumStateTransition.entries()) {
+      const dexOutput = await this.getDexOutput(transition.transactionHash)
+
+      const batch = await this.availabilityGatewayClient.getPerpetualBatch(
+        transition.batchId
+      )
+      const newPositions = this.buildNewPositions(batch)
+
+      await this.stateUpdater.processStateTransition(
+        {
+          id: id + i + 1,
+          blockNumber: transition.blockNumber,
+          stateTransitionHash: transition.stateTransitionFact,
+        },
+        dexOutput,
```

Comment:
You are still getting a typescript error here:
```
src/core/ValidiumSpotDataSyncService.ts(86,9): error TS2345: Argument of type 'StarkExDexOutput' is not assignable to parameter of type 'StarkExProgramOutput'.
 |   Type 'StarkExDexOutput' is missing the following properties from type 'StarkExProgramOutput': configurationHash, assetConfigHashes, oldState, newState, and 4 more.
```

Reply:
That's right. I just spoke to Adrian about this and I will create a PR with just the decodeDexOutput changes, without calling it here